mixer/pkg/runtime2/config/ephemeral.go

```diff
@@ -0,0 +1,383 @@
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming handler config: name='%s'\n%s", adapterName, resource.Spec.String())
+		}
+
+		cfg := &Handler{
+			Name:    adapterName,
+			Adapter: info,
+			Params:  resource.Spec,
+		}
+
+		handlers[cfg.Name] = cfg
+	}
+
+	counters.handlerConfig.Add(float64(len(handlers)))
+	return handlers
+}
+
+func (e *Ephemeral) processInstanceConfigs(counters Counters) map[string]*Instance {
+	instances := make(map[string]*Instance)
+
+	for key, resource := range e.entries {
+		var info *template.Info
+		var found bool
+		if info, found = e.templates[key.Kind]; !found {
+			continue
+		}
+
+		instanceName := key.String()
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming instance config: name='%s'\n%s", instanceName, resource.Spec.String())
+		}
+
+		cfg := &Instance{
+			Name:     instanceName,
+			Template: info,
+			Params:   resource.Spec,
+		}
+
+		instances[cfg.Name] = cfg
+	}
+
+	counters.instanceConfig.Add(float64(len(instances)))
+	return instances
+}
+
+func (e *Ephemeral) processRuleConfigs(
+	handlers map[string]*Handler,
+	instances map[string]*Instance,
+	counters Counters) []*Rule {
+
+	log.Debug("Begin processing rule configurations.")
+
+	var rules []*Rule
+
+	for ruleKey, resource := range e.entries {
+		if ruleKey.Kind != RulesKind {
+			continue
+		}
+		counters.ruleConfig.Add(1)
+
+		ruleName := ruleKey.String()
+
+		cfg := resource.Spec.(*config.Rule)
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming rule: name='%s'\n%s", ruleName, cfg.String())
+		}
+
+		// TODO(Issue #2139): resourceType is used for backwards compatibility with labels: [istio-protocol: tcp]
+		// Once that issue is resolved, the following block should be removed.
+		rt := resourceType(resource.Metadata.Labels)
+		if cfg.Match != "" {
+			m, err := expr.ExtractEQMatches(cfg.Match)
+			if err != nil {
+				log.Warnf("ConfigWarning: Unable to extract resource type from rule: name='%s'", ruleName)
+
+				// instead of skipping the rule, add it to the list. This ensures that the behavior will
+				// stay the same when this block is removed.
+			} else {
+				if ContextProtocolTCP == m[ContextProtocolAttributeName] {
+					rt.protocol = protocolTCP
+				}
+			}
+		}
+
+		// extract the set of actions from the rule, and the handlers they reference.
+		var actions []*Action
+		for i, a := range cfg.Actions {
+			log.Debugf("Processing action: %s[%d]", ruleName, i)
+
+			var found bool
+			var handler *Handler
+			handlerName := canonicalize(a.Handler, ruleKey.Namespace)
+			if handler, found = handlers[handlerName]; !found {
+				log.Warnf("handler not found: handler='%s', action='%s[%d]'",
+					handlerName, ruleName, i)
+				counters.ruleConfigError.Inc()
+				continue
+			}
+
+			// Keep track of unique instances, to avoid using the same instance multiple times within the same
+			// action
+			uniqueInstances := make(map[string]bool)
+
+			var actionInstances []*Instance
```

Comment:
`actionInstances = make([]*Instances, 0, len(a.Instances))`

needless allocations without this.